yt_dlp/extractor/twitcasting.py

```diff
@@ -79,74 +93,110 @@ def _real_extract(self, url):
-                get_element_by_class('tw-player-duration-time', webpage)))
+        duration = (try_get(video_js_data, lambda x: sum(float_or_none(y.get('duration')) for y in x) / 1000)
+                    or parse_duration(clean_html(get_element_by_class('tw-player-duration-time', webpage))))
         view_count = str_to_int(self._search_regex(
-            r'Total\s*:\s*([\d,]+)\s*Views', webpage, 'views', None))
+            (r'Total\s*:\s*([\d,]+)\s*Views', r'総視聴者\s*:\s*([\d,]+)\s*</'), webpage, 'views', None))
         timestamp = unified_timestamp(self._search_regex(
             r'data-toggle="true"[^>]+datetime="([^"]+)"',
             webpage, 'datetime', None))
 
-        if m3u8_url:
-            formats.extend(self._extract_m3u8_formats(
-                m3u8_url, video_id, 'mp4', 'm3u8_native', m3u8_id='hls', live=is_live))
-        self._sort_formats(formats)
+        stream_server_data = self._download_json(
+            'https://twitcasting.tv/streamserver.php?target=%s&mode=client' % uploader_id, video_id,
+            'Downloading live info', fatal=False)
 
-        return {
-            'id': video_id,
+        is_live = 'data-status="online"' in webpage or traverse_obj(stream_server_data, ('movie', 'live'))
+        if not traverse_obj(stream_server_data, 'llfmp4') and is_live:
+            self.raise_login_required(method='cookies')
+
+        base_dict = {
             'title': title,
             'description': description,
             'thumbnail': thumbnail,
             'timestamp': timestamp,
             'uploader_id': uploader_id,
             'duration': duration,
             'view_count': view_count,
-            'formats': formats,
             'is_live': is_live,
         }
 
+        def find_dmu(x):
+            data_movie_url = self._search_regex(
+                r'data-movie-url=(["\'])(?P<url>(?:(?!\1).)+)\1',
+                x, 'm3u8 url', group='url', default=None)
+            if data_movie_url:
+                return [data_movie_url]
+
+        m3u8_urls = (try_get(webpage, find_dmu, list)
+                     or traverse_obj(video_js_data, (..., 'source', 'url'))
+                     or [f'https://twitcasting.tv/{uploader_id}/metastream.m3u8'] if is_live else None)
+        if not m3u8_urls:
+            raise ExtractorError('Failed to get m3u8 playlist')
+
+        if is_live:
+            m3u8_url = m3u8_urls[0]
+            formats = self._extract_m3u8_formats(
+                m3u8_url, video_id, ext='mp4', m3u8_id='hls',
+                live=True, headers=self._M3U8_HEADERS)
+
+            try:
+                formats.extend(self._extract_m3u8_formats(
+                    m3u8_url, video_id, ext='mp4', m3u8_id='source',
+                    live=True, query={'mode': 'source'},
+                    note='Downloading source quality m3u8',
+                    headers=self._M3U8_HEADERS))
+            except ExtractorError as ex:
+                self.report_warning(ex)
```

Comment:
just use `fatal=False`?

Reply:
you're right
blaming myself at the time I wrote that